src/utils/actions.js

```diff
@@ -0,0 +1,325 @@
+          // code filter isn't under the same constraint as datefilter
+          // we can assume that it's just a path
+          const path = codeFilter.path;
+          const searchParam = codeFilter.searchParam;
+          if (path) {
+            // no need to refilter resources
+            filteredResources = filteredResources.filter(resource => {
+              const code = getByPath(resource, path);
+              // codeFilter.code is of type Coding, could also be a valueSet?
+              return compareCodes(code, codeFilter.code[0]);
+            });
+          } else if (searchParam) {
+            filteredResources = filteredResources.filter(resource => {
+              const code = getBySearchParam(resource, searchParam);
+              return compareCodes(code, codeFilter.code);
+            });
+          }
+        }
+        return context.records.push(...filteredResources);
+      });
+    }
+  },
+  'execute-research-query': context => {
+    // noop
+    return context;
+  }
+};
+
+function validateProfile(report, structureDefinition) {
+  // TODO: Perhaps this is something one of the servers/clients would handle?
+  if (structureDefinition) {
+    return report;
+  }
+}
+
+function getByPath(resource, path) {
+  const fhirpath = new FhirPath(resource);
+  return fhirpath.evaluate(path);
+}
+
+function compareCodes(coding1, coding2) {
+  // assume they're both coding
+  return coding1.code === coding2.code && coding1.system === coding2.system;
+}
+
+function getBySearchParam(resource, searchParam) {
+  // TODO: function to search resources by param
+  // and return correct attribute
+  return resource[searchParam];
+}
+
+function compareDates(date, period) {
+  // TODO: returns true if the periods overlap
+  const now = Date.parse(date);
+  const start = Date.parse(period.start);
+  const end = Date.parse(period.end);
+  return now > start && now < end;
+}
+
+function createBundle(records) {
+  const bundle = {
+    resourceType: 'Bundle',
+    type: 'message',
+    timestamp: Date.now(),
+    entry: []
+  };
+
+  records.forEach(record => {
+    bundle.entry.push({
+      resource: record
+    });
+  });
+
+  return bundle;
+}
+
+function evaluateExpression(expression, resource) {
+  if (expression.language === 'text/fhirpath') {
+    const fhirpath = new FhirPath(resource);
+    const path = fhirpath.evaluate(expression.expression);
+    return path;
+  }
+}
+
+function makeHeader(context) {
+  const header = {
+    resourceType: 'MessageHeader',
+    extension: [
+      {
+        url: 'http://hl7.org/fhir/us/medmorph/StructureDefinition/ext-messageProcessingCategory',
+        valueCode: 'consequence'
+      }
+    ],
+    eventCoding: {
+      system: 'http://hl7.org/fhir/us/medmorph/CodeSystem/us-ph-messageheader-message-types',
+      code: 'message-report'
+    },
+    destination: [
+      {
+        // assume client has some way to produce its endpoint
+        endpoint: context.client.dest.endpoint
```

Comment:
I can add this to the context since I am already pulling it out for the `context.client.dest.submit` operation. Lets make sure I collect all your changes. I requested your review on my PR

Reply:
the `testContext` resource has all the object elements needed for the actions, so that should be a good guide.